core/src/main/java/tachyon/client/TachyonFS.java

```diff
@@ -682,11 +674,14 @@ public synchronized String getUfsAddress() throws IOException {
    */
   @Override
   public synchronized TachyonURI getUri() {
-    String scheme = CommonConf.get().USE_ZOOKEEPER ? Constants.SCHEME_FT : Constants.SCHEME;
-    String authority = mMasterAddress.getHostName() + ":" + mMasterAddress.getPort();
-    return new TachyonURI(scheme, authority, TachyonURI.SEPARATOR);
+    if (mRootUri == null) {
```

Comment:
yes, move this block to constructor is safer. But I don't think this lazy solution will cause concurrency bug, since all fields referenced won't be changed by other methods.